doc/src/lexer_tutorial/004_external_lib.md

```diff
@@ -0,0 +1,330 @@
+[crates.io](https://crates.io) many different libraries to generate a lexer for
+you.
+
+In this tutorial, we will use [Logos](https://docs.rs/logos/latest/logos/) to
+build a simple lexer for a toy programming language. Here is an example of what
+we will be able to parse:
+
+```
+var a = 42;
+var b = 23;
+
+# a comment
+print (a - b);
+```
+
+## Setup
+
+In your `Cargo.toml`, add the following dependency:
+
+```toml
+logos = "0.12.0"
+```
+
+This will provide the `logos` crate and the `Logos` trait.
+
+## The AST
+
+As an example, we will use this AST in our grammar:
+
+```rust
+#[derive(Clone, Debug, PartialEq)]
+pub enum Statement {
+  Variable { name: String, value: Box<Expression> },
+  Print { value: Box<Expression> },
+}
+
+#[derive(Clone, Debug, PartialEq)]
+pub enum Expression {
+  Integer(i64),
+  Variable(String),
+  BinaryOperation { lhs: Box<Expression>, operator: Operator, rhs: Box<Expression> },
+}
+
+#[derive(Clone, Debug, PartialEq)]
+pub enum Operator {
+  Add,
+  Sub,
+  Mul,
+  Div,
+}
+```
+
+## Implement the tokenizer
+
+In a file named `tokens.rs` (or any other name you want), create an enumeration
+for your tokens:
+
+```rust
+use std::fmt;  // to implement the Display trait
+use logos::Logos;
+
+#[derive(Logos, clone, Debug, PartialEq)]
+pub enum Token {
+  #[token("var")]
+  KeywordVar,
+  #[token("print")]
+  KeywordPrint,
+
+  #[regex("[_a-zA-Z][_0-9a-zA-Z]*", |lex| lex.slice().parse())]
+  Identifier(String),
+  #[regex("\d+", |lex| lex.slice().parse())]
+  Integer(i64),
+
+  #[token("(")]
+  LParen,
+  #[token(")")]
+  RParen,
+  #[token("=")]
+  Assign,
+  #[token(";")]
+  Semicolon,
+
+  #[token("+")]
+  OperatorAdd,
+  #[token("-")]
+  OperatorSub,
+  #[token("*")]
+  OperatorMul,
+  #[token("/")]
+  OperatorDiv,
+
+  #[regex(r"#.*\n?", logos::skip)]
+	#[regex(r"[ \t\n\f]+", logos::skip)]
+  #[error]
+  Error,
+}
+```
+
+The annotation `#[token()]` will generate the enum's variant when the token is
+detected in the input string.
+
+The annotation `#[regex()]` will generate the enum's variant when the regex
+match in the input string.
+
+A few things to note about how **Logos** works:
+
+ - if a `#[regex()]` can match the same token as a `#[token()]`, the later will
+   have priority
+ - longer tokens have priority
+ - the `#[error]` annotation is required, this variant will be generated for
+   patterns that did not match any rules, we use this rule to skip comments and
+   whitespaces
+
+This means the `"printa"` input string will generate the following token:
+
+ - `Token::Identifier(String::new("printa"))`
+
+And not:
+
+ - `Token::KeywordPrint`
+ - `Token::Identifier(String::new("a"))`
+
+This is because `printa` is longer than `print`, therefore the `Identifier` rule
+has priority.
+
+Finally, we implement the `Display` trait:
+
+```rust
+impl fmt::Display for Token {
+  fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+    write!(f, "{:?}", self)
+  }
+}
+```
```

Comment:
The Token is part of the error message generated by LALRPOP in case of grammatical errors. Therefore it needs the `Display` trait if you don't want to `.unwrap()`.

Maybe just a warning is enough?